dask_kubernetes/operator/controller/controller.py

```diff
@@ -392,67 +403,92 @@ async def get_desired_workers(scheduler_service_name, namespace, logger):
         ) from e
 
 
-@kopf.on.update("daskworkergroup.kubernetes.dask.org")
-async def daskworkergroup_update(spec, name, namespace, logger, **kwargs):
-    async with kubernetes.client.api_client.ApiClient() as api_client:
-        customobjectsapi = kubernetes.client.CustomObjectsApi(api_client)
-        corev1api = kubernetes.client.CoreV1Api(api_client)
+worker_group_scale_locks = defaultdict(lambda: asyncio.Lock())
```

Comment:
Adding a lock like this makes a lot of sense. However, this does prevent the controller from scaling. What happens if there are two or more instances of the controller? They will have different locks.

We may need to think of some other way to ensure that only one scale call can be running at a time.

Reply:
(this is a response to both this comment and the HA portion of your top level review comment)

Yeah I had a feeling tracking the locks internally wasn't the kopf way, it certainly felt out of place when I was writing. I've just done some more digging into this and I wasn't able to find any information on HA kopf deployments, can you point me to what you're referencing? I did notice a couple other things while digging in:

- I came across [this example](https://github.com/nolar/kopf/blob/5c6f68c9dfd80a23dc54bc50a951c3c2535476f3/examples/13-hooks/example.py) which seems to be using a similar patter to us, so at least in a non HA setup we don't seem to be too far off?
- I was reading the [peering docs](https://github.com/nolar/kopf/blob/5c6f68c9dfd80a23dc54bc50a951c3c2535476f3/examples/13-hooks/example.py) and noticed in the section on priority they discuss ensuring two operators aren't activate at the same time but instead ensure that one is on standby ready to pick up if the other drops. I'm wondering if this is the HA setup you were referencing?
- I noticed [this line](https://github.com/nolar/kopf/blob/5c6f68c9dfd80a23dc54bc50a951c3c2535476f3/examples/13-hooks/example.py) in the peering docs: Usually, one and only one operator instance should be deployed for the resource. If that operator’s pod dies, the handling of the resource of this type will stop until the operator’s pod is restarted (and if restarted at all). Which reads to me as if they intend only one operator instance to be running for a given resource.

Given the above, I'm wondering if our current setup is done in the way kopf intends? I'll defer to you here as I'm brand new to k8s operators and kopf but calling out for discussion!

Reply:
Thanks for the response. My assumption was that multiple instances of the controller would divide up events, but it looks like that assumption was wrong and the behaviour is more of an active-backup model which works in our favour here.

My worry is still valid because if it fails over to the backup controller, the locks will be lost. Then if it fails back the other way there will be old and potentially out-of-date locks.

There is also an assumption that the controller can be safely restarted at any time. So it shouldn't really be holding things like locks in memory.

I think my suggestion would be to use a Kubernetes resource to manage this state. Looking at [these docs](https://kopf.readthedocs.io/en/stable/continuity/) it seems the peering code uses annotations on a `KopfPeering` resource to track state. Given that we are tracking locks for scaling a `DaskWorkerGroup` perhaps we should be storing the lock in an annotation on that object?

Reply:
Yup I think a kopf based annotation lock would work well.  I'll include some thoughts on how we should implement it here and I've also included some additional thoughts on the pure memory lock vs annotation lock.  Once we decide I'll implement and update!

Starting with thoughts on the method:
- The key question is what do we need to ensure is locked across function executions and processes.  In the case of `daskworkergroup_replica_update` we need to ensure that the (inclusive) block between calculating `current_workers` and making the API calls to k8s is executed serially.  We need to make sure that between the time we read the state of the cluster and apply changes, nothing else comes along and tries to read the current state.  
- I'm assuming that we can rely on kopf to ensure that only one operator is running at once, so a new operator wont start operating until the others are down.  If that isn't true then I think we need to go with the annotation approach.
- If the above assumption is true, then I think we can be sure that when the second operator starts running, the first one has stopped and thus the lock will be either cleanly or forcefully released by the first one.
  - If the lock is cleanly released then the scaling process completed (or at least finished in a steady state) and the next operator is free to pick it up
  - If the lock is forcefully released, the scaling process may be left in a intermittent state (half the pods terminated, half the pods created, etc).  This intermittent state is captured in the state of the pods and the operator is required to read the state of the pods (terminating, creating, etc) when it is determining what pods to create / destroy.  The operator already does this to make itself resilient to pod lifecycles states / changes. 
  - So in the case of this method I think the critical parts that must be executed serially are in fact executed serially, I can't think of a way to get the critical block to execute in parallel

Thoughts on how we can setup the annotation based approach:
- Use a UID for the operator instance (pull from kopf or generate one ourselve)
- Use two annotations (`owner_uid` and `heartbeat_timestamp`)
- To acquire the lock (following a similar pattern to distributed locking with DynamoDB / Redis):
  - Check if a lock exists
  - If a lock exists, wait for the lock to be released by looping to check, if the lock heartbeat goes stale for some period of time (ie the last time the `heartbeat_timestamp` is not updated for some number of minutes) then forcefully acquire the lock (this prevents a deadlock from a force kill of the operator)
  - Once the lock does not exist or we decide to force acquire, set the annotation of owner and timestamp and start an async timer that updates the heartbeat annotation every 10 seconds.
    - This needs to be done in an atomic "get / set" method  which essentially asserts that nobody has written the lock in parallel.  It ensures that if two processes are trying to acquire the same lock, only one wins.  In Dynamo I believe this is done with a condition expression that atomically ensures no lock record exists before writing the lock.  If another parallel write wins out, then the condition expression will fail the write and the lock won't be acquired.  I did some digging into the k8s API and it looks like if we include a resource version in the metadata and we use a replace call instead of an update (patch), the API server will reject the change if something else has written.
- To release the lock, delete the annotations
- Manage this with an async context manager as suggested

I think the timestamp / force acquire approach is necessary to ensure we don't step on another operator that is shutting down / finishing work before releasing the lock.  Without the timestamp, the new operator would need to forcefully acquire the lock when it starts up, which would mean we have a situation where the first operator could still be running thinking it has the lock.   As mentioned above, if we can rely on kopf to ensure that the new operator's callbacks don't run until the old one is shut down then I think this isn't necessary.  If we can't rely on kopf and forcefully acquire without the timestamp, there's a short time where two scaling processes could be executing together for the same worker group.  If we can rely on the operator we could still make an argument for future proofing things to account for logic that needs to be locked across processes, but I'm not sure the ROI on it is there.

Let me know what you think and I'll update!

Reply:
> As mentioned above, if we can rely on kopf to ensure that the new operator's callbacks don't run until the old one is shut down then I think this isn't necessary.

My reading of the kopf peering docs makes me think we could rely on this.

I'll leave it to you how you want to implement it. I would be happy with either approach. If we agree that we can rely on kopf my advice would be to keep it as simple as possible.

Reply:
I've given them another read and I'm in agreement we can rely on kopf so I suggest we stick with the current (simpler) implementation 👍